AppCenterDataStorage/AppCenterDataStorage/MSDataSourceError.m

```diff
@@ -2,16 +2,70 @@
 // Licensed under the MIT License.
 
 #import "MSDataSourceError.h"
+#import "MSDataStoreErrors.h"
+
+/**
+ * CosmosDb Http code key.
+ */
+static NSString *const kMSCosmosDbHttpCodeKey = @"com.Microsoft.AppCenter.HttpCodeKey";
 
 @implementation MSDataSourceError
 
 @synthesize error = _error;
+@synthesize errorCode = _errorCode;
 
 - (instancetype)initWithError:(NSError *)error {
+  return [self initWithError:error errorCode:[MSDataSourceError errorCodeWithError:error]];
+}
+- (instancetype)initWithError:(NSError *)error errorCode:(NSInteger)errorCode {
   if ((self = [super init])) {
     _error = error;
+    _errorCode = errorCode;
   }
   return self;
 }
 
++ (NSInteger)errorCodeWithError:(NSError *)error {
+  if (error == nil) {
+    return kMSACDocumentSucceededErrorCode;
+  }
+
+  // Get error code.
+  NSNumber *errorCode = [error userInfo][kMSCosmosDbHttpCodeKey];
+  switch ([errorCode integerValue]) {
```

Comment:
What is this mapping? Can the int itself just be returned?

Reply:
the user info is defined as `@property(readonly, copy) NSDictionary<NSErrorUserInfoKey, id> *userInfo;` so it's really up to us. I don't think there's an effect in practice. From a consistency standpoint, we don't have a large enough dataset in our SDK ;) there's only a handful of cases where we return an `int` or a `NSInteger`.

Reply:
user will be able to use this enumeration to identify errors. it's also good practice to provide custom message along with this error code(as opposed to generic error message coming from CosmosDB)

Reply:
Perhaps I am missing something, or I wasn't specific enough. The return type of this method is `NSInteger`. We do a switch on the int, and return an enum with the same value as the code we already have, and the enum type is `NSInteger`. Further, from the callee's point of view, the return type is an `NSInteger`. So there would be no perceivable difference if we just returned `[errorCode integerValue]`. So why do we need to switch? I am under the impression that this essentially is saying `case 200: return 200`, etc.

Reply:
Discussed offline; I didn't realize it was a common practice to make sure that all the error codes are handled.